### DISCUSSION TITLE
Holiday lights scheduler and demo mode for WLED

### DISCUSSION THREAD

**Discussion by zachowj** [zachowj/node-red-contrib-home-assistant-websocket#306] [Cookbook]
Here's a flow that's been to control some outdoor house LEDs hooked up to an ESP8266 running WLED.

The holiday lights turn on each day at sunset and then off at midnight. It uses WLED presets based on the day or season. It has a demo mode that will either cycle through all the defined presets or all the included effects.

![image](https://user-images.githubusercontent.com/37859597/102045159-37284080-3d8d-11eb-8653-ea3ac86425d1.png)

A write up and the JSON export can be found in the [cookbook section](https://zachowj.github.io/node-red-contrib-home-assistant-websocket/cookbook/holiday-lights-scheduler-and-demo-mode-for-wled.html) of the documentation.

**Comment by RestNPizza** [comment]
Hello,

First, i'd like to say thank you very much for creating this flow. I've implemented it in my own home assistance/node red instance and love it. After going through the flow to understand it better, I was able to alter it to fit my use case. There was one thing I am unsure about and wanted to get more info on. Let me preface this by saying I'm a complete novice with javascript. Could you break down in basic terms the following line that is found in the "date ranges" function node and what it is used for?

const secondOutput = updateInterval ?  { payload: { timeout: updateInterval } } : { reset: true };

Thank you!!

**Reply by zachowj** [reply]
For most date ranges a preset is triggered at sunset and runs until midnight. But for some date ranges you may want to have multiple presets that are randomly chosen for that same night. In the exported flow this is done for Christmas time where there are 4 different presets that it randomly selects every 60 minutes.

`updateInterval` is just that, how often you want to change the preset.

<https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Conditional_Operator>

The statement you are asking about is checking if `updateInterval` is true. In javascript when a variable is a number, it is false when it equals zero and true when it does not. 

So if it is greater than zero it will send a payload to the second output of that function node triggering the wait-until to wait for that interval time and then send input back to the function node to randomly selected a new preset or if the light entity is turned off while waiting it just ends the flow.

If `updateInterval` is 0 send a `msg` with the property `reset` set to `true`. Basically, this does nothing other than resetting the wait-until node if it was waiting and stops the flow.

Line 4-9
```js
const christmasPresets = [
  "Christmas",
  "Christmas Classic",
  "Christmas Holly Jolly",
  "Christmas Candy Cane"
];
```

Line 13 - default interval is set to zero
```js
let updateInterval = 0; // in minutes
```

Line 55-59 - if it this date range set interval to 60 minutes 
```js
    case isToday("Thanksgiving Day", 1231):
    case isToday(101, 106):
      preset = getRandomPreset(christmasPresets);
      updateInterval = 60;
      break;
```

**Reply by RestNPizza** [reply]
Awesome explanation! Thanks so much!

For some reason, despite the updateInterval being set to 0 in my situation, it is still registering as true. This is causing the message that comes out of this function node to be the reset message. For some reason, the msg.payload that is supposed to display the id of the preset that matches the date conditions doesn't come through. Its like the secondOutput message is overriding the msg.payload set to the preset. Not sure why this could be. I didnt alter any of the code other than unique identifiers applicable to my home assistant server (the flow get commands and states command). The node correctly displays the correct preset name under it when executing (in my current case, it displays "Spring") but the reset: true message is the only thing that comes out of it.

Here's my code:
```
const now = new Date();
const monthDay = getMonthDay(now);
const { idLookup, nameLookup } = flow.get("wledDeskLedsPresets");
const christmasPresets = [
  "Christmas",
  "Christmas Classic",
  "Christmas Holly Jolly",
  "Christmas Candy Cane"
];
const states = global.get("homeassistant").apolloHomeAssistant.states;
const entityId = flow.get("wledDeskLedsEntityId");

let updateInterval = 0; // in minutes
let preset;

switch(true) {
// Winter 
    case isToday(214):
      preset = "Valentine";
      break;
// President's Day changes every year, Veteran's Day & Flag Day always the same date
    case isToday(614): // 6/14 Flag Day
    case isToday(1111): // 11/11 Veterans Day
    case isToday("President's Day"):
    case isToday("Memorial Day"):
      preset = "Red, White, and Blue";
      break;
    case isToday(317):
      preset = "St Patty";
      break;
    case isToday("Easter"):
      preset = "Easter";
      break;
// Spring
    case isToday(504):
      preset = "Starwars";
      break;
    case isToday(505):
      preset = "Cinco de Mayo";
      break;
// Summer
    case isToday(704):
      preset = "Independence Day";
      break;
// Autumn
    case isToday(1017, 1031):
      preset = "Halloween";
      break;
    case isToday("Thanksgiving Day"):
      preset = "Autumn";
      break;
    case isToday(1231):
      preset = "New Years";
      break;
    case isToday("Thanksgiving Day", 1231):
    case isToday(101, 106):
      preset = getRandomPreset(christmasPresets);
      updateInterval = 60;
      break;
    default:
      const seasons = {
        "winter": "Winter",
        "spring": "Spring",
        "summer": "Summer",
        "autumn": "Autumn"
      }
      const currentSeason = states["sensor.season"].state;
      preset = seasons[currentSeason];
      break;
}

node.status({ text: preset});
msg.payload = nameLookup[preset];
const secondOutput = updateInterval ?  { payload: { timeout: updateInterval } } : { reset: true };

return [msg, secondOutput];

function isToday(start, end) {
  if (typeof start === "string") {
    start = getHolidayDate(now, start)
  }
  if (typeof end === "string") {
    end = getHolidayDate(now, end)
  }
  if (end === undefined) return monthDay === start;

  return monthDay >= start && monthDay <= end;
}

function getHolidayDate(date, holiday) {
    holiday = holiday.toLowerCase();
    // dates are formatted as month, week, day of the week
    // names should be lowercase
    const holidays = {
        "martin luther king day": [0,2,1],
        "president's day": [1,2,1],
        "daylight savings time begins": [2,1,0],
        "mother's day": [4,1,0],
        "memorial day": [4,-1,1],
        "father's day": [5,2,0],
        "labor day": [8,0,1],
        "columbus day": [9,1,1],
        "daylight savings time ends": [10,0,0],
        "thanksgiving day": [10,3,4]
    };
    let holidayDate;
    if(holiday === "easter") {
        holidayDate = getEasterDate(date); 
    } else {
        const [month, week, day] = holidays[holiday];
        holidayDate = getDate(date.getFullYear(), month, week, day);
    }
    
    const monthDay = getMonthDay(holidayDate);
    
    return monthDay;
}

function getMonthDay(date) {
    const monthString = String(date.getMonth() + 1).padStart(2, "0");
    const dayString = String(date.getDate()).padStart(2, "0");
    const monthDay = Number(`${monthString}${dayString}`);
    
    return monthDay;
}

function getDate(year, month, week, day) {
    let firstDay = 1;
    if (week < 0) {
        month++;
        firstDay--;
    }
    const date = new Date(year, month, (week * 7) + firstDay);
    if (day < date.getDay()) {
        day += 7;
    }
    date.setDate(date.getDate() - date.getDay() + day);
    return date;
}

/**
 * Calculates Easter in the Gregorian/Western (Catholic and Protestant) calendar
 * based on the algorithm by Oudin (1940) from http://www.tondering.dk/claus/cal/easter.php
 * @returns {array} [int month, int day]
 */
function getEasterDate(date) {
  const year = date.getFullYear();
  const f = Math.floor,
    // Golden Number - 1
    G = year % 19,
    C = f(year / 100),
    // related to Epact
    H = (C - f(C / 4) - f((8 * C + 13) / 25) + 19 * G + 15) % 30,
    // number of days from 21 March to the Paschal full moon
    I = H - f(H / 28) * (1 - f(29 / (H + 1)) * f((21 - G) / 11)),
    // weekday for the Paschal full moon
    J = (year + f(year / 4) + I + 2 - C + f(C / 4)) % 7,
    // number of days from 21 March to the Sunday on or before the Paschal full moon
    L = I - J,
    month = 3 + f((L + 40) / 44),
    day = L + 28 - 31 * f(month / 4);

  return new Date(`${month}/${day}/${year}`);
}

function getRandomPreset(list) {
  const activePreset = states[entityId].attributes.preset;
  const presetName = idLookup[activePreset];
  const filteredList = list.filter((e) => e !== presetName);
  
  return filteredList[Math.floor(Math.random() * filteredList.length)];
}
```


One more thing...is there any way for you to share your wled presets? Per the date ranges node, it looks like you have quite a few and i'm really curious what you've come up with.

Thanks again for this awesome setup!

**Reply by RestNPizza** [reply]
Ok, i'm an idiot. I didn't realize I was feeding everything off the bottom output of the node. This is why only the secondOutput was coming through. I fixed it and everything is working beautifully! Thanks so much for your help!

Per the last question on my previous reply, would you be able to share your wled presets? Per the date ranges node, it looks like you have quite a few and i'm really curious what you've come up with.

Thanks again!

**Reply by zachowj** [reply]
https://gist.github.com/zachowj/3bf93f3cbcc94ade93ae27c33accff12

**Comment by pvr2002** [comment]
Appreciate you sharing this with the community.  So far everything seems great.  Would you be able to share your WLED presets for each of the holidays as well?

**Reply by zachowj** [reply]
https://gist.github.com/zachowj/3bf93f3cbcc94ade93ae27c33accff12

**Reply by RestNPizza** [reply]
Thank you for sharing. Is there an easy way to import preset? I couldn't find anything online anywhere that shows this. Thanks again!

**Reply by zachowj** [reply]
Append `/edit` to the WLED url and you can edit and save the preset.json file. You'll need to reboot the device after saving for the presets to show up.

```
http://wledip/edit
````

**Reply by RestNPizza** [reply]
Awesome! Thanks so much!!!

**Comment by sys1aeg** [comment]
Hi - thank you for providing this it is truly amazing 
having issues getting presets to work - from everything i looked at presets need to be a numeric value 
how are setting the the preset name value ?

**Reply by zachowj** [reply]
The newer versions of HA now expose presets as a `select` entity in HA. I've updated this on my setup but need to update the post to take a advantage of this.

The `process data` node gets the list of presets and builds a key/value object based on their name. Then in `date ranges` it uses that table to output the correct preset id.

**Reply by sys1aeg** [reply]
Thank you so much for your quick response !

**Comment by sys1aeg** [comment]
so function getHolidayDate  table values are zero relative ?

**Reply by zachowj** [reply]
Yes, javascript months and weekdays are zero-based.

**Comment by sys1aeg** [comment]
Hi 
 trying to get demo mode to work in function get next preset . the cycle effect are working 
but cycle presets are not. looks like statement:
const activePreset = states[entityId].attributes.preset; is unkown 
which element lookup always returns a -1
any ideas would be appreciated 
thanks !

**Reply by zachowj** [reply]
In the newest version of HA/WLED the presets are now broken into a separate `select` entity.

try

```
const activePreset = states["select.wled_preset"].state;
```

with `select.wled_preset` being the entity id of the select

**Reply by sys1aeg** [reply]
Thanks you for that, yes that did work for activePreset 
but still getting a -1 from element using keys array.
added array Presetlist but not getting all presets 
here is my code- maybe using two arrays is not a good idea 

const states = global.get("homeassistant").homeAssistant.states;
const entityId = flow.get("wledHolidayLightsEntityId");
const { idLookup, nameLookup } = flow.get("wledHolidayLightsPresets");
const keys = Object.keys(idLookup);
//const activePreset = states[entityId].attributes.preset;
const Presetlist = states["input_select.holiday_lights_presets"].attributes.options;
const activePreset = states["select.wled_preset_4"].state;
let index = Presetlist.findIndex(ele => ele == activePreset);
index = index === -1 || index >= keys.length - 1 ? 0 : index +1;
const name = idLookup[keys[index]];

node.status({text: name });

msg.preset = keys[index];
msg.name = name;
msg.states = states;
return msg;

**Reply by zachowj** [reply]
I updated the cookbook example. Give that a try.

**Reply by sys1aeg** [reply]
Yes !
Works great !
Thank you very much --
now I have holiday lights up all year long thanks 
to you and Drzzz!

**Reply by sys1aeg** [reply]
Hi 
  looks like the latest release of HA has removed the wled.preset 
and I am trying the native call change to node "turn on holiday lights"
without success. thing were working great until the last update.

**Reply by sys1aeg** [reply]
ok i think i got it, changed node "date ranges" to pass payload name not number
and also changed "turn on holiday lights " to use  select.wled_preset
happy holidays !

**Comment by Drewskieza** [comment]
Good day,
Thanks for this.
Just two things:
1. With the new version of WLED there is now playlists (not sure how to integrate this into the system)
2. I have a automation that sets the same set of lights to a different preset when my front gate opens and turns the lights off once the gate closes.  I am trying to figure out a way to trigger this schedule again once the gate closes so that it goes back to the correct playlist if it still falls within the eztimer period. Any guidance for the technologically challenged will be appreciated.

**Reply by zachowj** [reply]
> 1. With the new version of WLED there is now playlists (not sure how to integrate this into the system)

This is assuming you want to use preset and playlist together. Playlists are exposed to HA as a `select` the same as presets. You can add a boolean to show if the current date is a playlist and then override the output to the call-service node to use the entity of the playlist entity id.

In the `date range` function node

Line 15
```
let updateInterval = 0; // in minutes
let preset;
let playlist = false; 
```

in a date block
```
    case isToday(317):
      preset = "St Patty";
      playlist = true;
      break;
```

Line 77
```
msg.payload = playlist ? { data: {entity_id: select.playlistentityId, option: preset}} : preset;
```



> 2\. I have a automation that sets the same set of lights to a different preset when my front gate opens and turns the lights off once the gate closes.  I am trying to figure out a way to trigger this schedule again once the gate closes so that it goes back to the correct playlist if it still falls within the eztimer period. Any guidance for the technologically challenged will be appreciated.

Add an event-state node for when the gate closes and run it into the "dark" switch node

**Comment by JHClagg** [comment]
Thanks so much for this awesome tool! I'm having a little issue, I've followed the instructions for setup and can get the lights to run via the "Manual Start" button, however I'm having an issue with the time-based starts. When I deploy I get an error that says "Call-Service attempted without connection to server." The EzTimer node show that it turns on but flow.dark is true doesn't seem to make its way to the other nodes in the flow. Has anyone else experienced this?

**Reply by zachowj** [reply]
The timer node triggers the flow on deployment and you probably have your deployment type set to full. So NR has to reconnect to HA every time you deploy which normally takes a few seconds. The call-service node is triggered before NR has connected to HA is the reason you see the error.

Edit the call-service node and enable queueing

**Reply by JHClagg** [reply]
Thanks, I have queueing set to "All Messages" but am still having this issue. It's strange because I have another routine setup using the Eztimer node to turn on the power supply for my lights with a flow and it works fine.

**Comment by jmsblah** [comment]
I am getting an error in the "getEasterDate" function in the "date ranges" node:
`Type 'Date' is missing the following properties from type 'any[]': length, pop, push, concat, and 31 more.(2740)`

**Reply by zachowj** [reply]
That is just a type error. It doesn't stop it from working. You can change line 150 and the error should go away.

from 

```
 * @returns {array} [int month, int day]
```

to 

```
 * @returns Date
```

**Comment by cmse-sycotix** [comment]
I know this thread has not had much activity in a while. My question is where do the light.holiday_lights and select.holiday_lights_presets come from? I followed the cookbook and homeassistant shows these entities as not available.\
-thanks

**Reply by zachowj** [reply]
Those are the entities exposed by your WLED device in Home Assistant.
